feat: remove bincode support, use msgpack-compact only

ludamad opened the pull request:
This moves us over to the msgpack-compact serialization format for ACIR. As well, the coming changes will no longer have msgpack_pack generated for these objects, as it was code only used in tests. I have removed such tests. The bbapi tests that use this are nice but would need to be moved to tests that consume acir artifacts. but most were quite simple.

# Summary
- Replaced `deserialize_any_format` with `deserialize_msgpack_compact`, moving everything to new msgpack format
- Updated test files to use msgpack serialization
- Removed bincode declarations and implementations from serde files

Fixes A-450

aakoshh commented on barretenberg/cpp/scripts/msgpack_compact_filter.py:
```diff
@@ -0,0 +1,243 @@
+        r'throw_or_abort\("expected MAP or ARRAY for " \+ name\);\s*\}\s*'
+        r'\}',
+        re.DOTALL
+    )
```
These seem like they can easily break if we change anything in the `Helpers`, so I wouldn't rely on this Python script in the future, and would rather make it configurable in the producer, but good on you for going the extra mile!

aakoshh commented on barretenberg/cpp/scripts/msgpack_compact_filter.py:
```diff
@@ -0,0 +1,243 @@
+            [&packer, tag](const auto& arg) {
+                packer.pack_map(1);
+                packer.pack(tag);
+                arg.msgpack_pack(packer);
```
That's nice, we should port it back to the code generator.

ludamad committed:
feat: remove bincode support, use msgpack-compact only

This moves us over to the msgpack-compact serialization format for ACIR. As well, the coming changes will no longer have msgpack_pack generated for these objects, as it was code only used in tests. I have removed such tests. The bbapi tests that use this are nice but would need to be moved to tests that consume acir artifacts. but most were quite simple.

# Summary
- Replaced `deserialize_any_format` with `deserialize_msgpack_compact`, moving everything to new msgpack format
- Updated test files to use msgpack serialization
- Removed bincode declarations and implementations from serde files

Fixes A-450

Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.cpp:
```diff
@@ -271,52 +271,29 @@ void update_max_witness_index_from_opcode(Acir::Opcode const& opcode, AcirFormat
+    BB_ASSERT(!buf.empty(), "deserialize_msgpack_compact: buffer is empty");
+
+    // Expect format marker for msgpack or msgpack-compact
+    const uint8_t FORMAT_MSGPACK = 2;
```
Did this start at 2 because it was implicitly the 2nd option after bincode?

ledwards2225 replied:
Maybe we don't want to change anyway to reduce breaking-ness?
